Reduce RPC branches for Python/BuiltinOp/TorchScript

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32753 Remove Python dependency in rref_impl.h/cpp
* **#32689 Reduce RPC branches for Python/BuiltinOp/TorchScript**

As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

xush6528 committed:
Reduce RPC branches for Python/BuiltinOp/TorchScript

As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 27181bf6:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=8e4161517e112478a1c1f0290fedb91965f95aff&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 8e416151 since Jan 28
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=27181bf6c3453ac6b0d1eb2edaa9942a2209ecea).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_xenial_py3_6_gcc5_4_test](https://dr.pytorch.org/build-details.html?build_id=91837202) from Jan 28



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xush6528 committed:
Update on "Reduce RPC branches for Python/BuiltinOp/TorchScript"


As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

[ghstack-poisoned]

xush6528 committed:
Update on "Reduce RPC branches for Python/BuiltinOp/TorchScript"


As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

[ghstack-poisoned]

xush6528 committed:
Update on "Reduce RPC branches for Python/BuiltinOp/TorchScript"


As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

[ghstack-poisoned]

xush6528 committed:
Update on "Reduce RPC branches for Python/BuiltinOp/TorchScript"


As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

[ghstack-poisoned]

xush6528 committed:
Update on "Reduce RPC branches for Python/BuiltinOp/TorchScript"


As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

[ghstack-poisoned]

xush6528 committed:
Update on "Reduce RPC branches for Python/BuiltinOp/TorchScript"


As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

[ghstack-poisoned]

xush6528 committed:
Update on "Reduce RPC branches for Python/BuiltinOp/TorchScript"


As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -833,8 +828,11 @@ def test_py_raise_in_user_func(self):
     def test_script_function_exception(self):
-        dst_rank = (self.rank + 1) % self.world_size
-
+        @torch.jit.script
```
torchscript functions does no need to be top-level ones? Because they are looked up in cu instead of using interpreter?

xush6528 replied:
I think so. `@torch.jit.script` decorator register them in `cu`. So Python global name lookup requirment does not apply here.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -336,8 +336,8 @@ If the future completes with an error, an exception is thrown.
   module.def(
-      "_invoke_rpc_script",
-      [](const std::string& dst,
+      "_invoke_rpc_torchscript",
```
thanks for fixing the name

mrshenli commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -12,14 +12,16 @@ namespace distributed {
 namespace rpc {
```
what about the file name here? do we need to keep it consistent as `torchscript_functions.h/cpp`?

mrshenli reviewed (approved):
LGTM! 

Looks like current master is in better shape now. Could you please rebase to that (or d119de8 which is in better shape:) ) and relaunch tests? Too many tests failed due to irrelevant things.

xush6528 committed:
Update on "Reduce RPC branches for Python/BuiltinOp/TorchScript"


As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

[ghstack-poisoned]

xush6528 committed:
Update on "Reduce RPC branches for Python/BuiltinOp/TorchScript"


As described in https://github.com/pytorch/pytorch/issues/32565

Differential Revision: [D5721814](https://our.internmc.facebook.com/intern/diff/D5721814/)

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/rpc/api.py:
```diff
@@ -449,19 +446,23 @@ def _invoke_rpc(to, func, rpc_type, args=None, kwargs=None):
+        fut = _invoke_rpc_builtin(dst_worker_info, qualified_name, rf, *args, **kwargs)
+    elif isinstance(func, torch.jit.ScriptFunction):
+        fut = _invoke_rpc_torchscript(
+            dst_worker_info.name, torch._jit_internal._qualified_name(func), *args, **kwargs
```
are you planning to pass "rf" (autograd profiler) as well in this diff or follow up diff?

xush6528 replied:
Yes. With this streamlining, it's easier to extend profiler to the new code path.

zhaojuanmao reviewed (approved):
lgtm, thanks!

xush6528 commented:
The 2 failures are irrelevant. 

pytorch_linux_xenial_py2_7_9_test:
test_jit_string.py - TestScript. test_str_ops

pytorch_linux_backward_compatibility_check_test:
quantized::linear_relu_dynamic_fp16(Tensor X, Tensor W_prepack) -> (Tensor Y)
